### PR TITLE
Add NEWS and docs for the new inlining algorithm

### PR DESCRIPTION
**Pull request by timholy**
I also included a number of useful tips I've accumulated while hacking on `inference.jl`. Hopefully this will save head-scratching for future generations. (Others are encouraged to add to this!)

[ci skip]

### CODE REVIEW

File: doc/src/devdocs/inference.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+[standard ranges for common architectures](http://ithare.com/wp-content/uploads/part101_infographics_v08.png)
+(see
+[Agner Fog's analysis](http://www.agner.org/optimize/instruction_tables.pdf)
+for more detail).
+
+We supplement this low-level lookup table with a number of special
+cases. For example, an `:invoke` expression (a call for which all
+input and output types were inferred in advance) is assigned a fixed
+cost (currently 20 cycles). In contrast, a `:call` expression, for
+other than intrinsics/builtins, indicates that the call will require
```
Comment (tkelman) [inline doc/src/devdocs/inference.md]: for functions other than ?

**Review by vtjnash — APPROVED** [review]

File: doc/src/devdocs/inference.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+
+# Inference
```
Comment (vtjnash) [inline doc/src/devdocs/inference.md]: Perhaps we should add links to https://juliacomputing.com/blog/2016/04/04/inference-convergence.html and https://juliacomputing.com/blog/2017/05/15/inference-converage2.html to describe the overall structure?

I should also probably add a description of https://github.com/JuliaLang/julia/pull/21933 here too, later, when I get a chance.

**Comment by timholy** [comment]
Hmm, link is broken. Any idea what I missed?

**Comment by timholy** [comment]
I see there's a blank line at the top of `inference.md`. Could that be it? (I am not aware of a way to test without merging to master, so I thought it would be worth asking if anyone knows.)

**Comment by timholy** [comment]
Fixed in 9c1d1b8ba8432463d752af27b0ea6257187efc39

**Comment by tkelman** [comment]
`make docs` ?

**Pull request merged**